lib/octopus_serverspec_extensions/type/octopus_deploy_upgrade_config.rb

```diff
@@ -0,0 +1,94 @@
+require 'serverspec/type/base'
+require 'net/http'
+require 'json'
+
+module Serverspec::Type
+  class OctopusDeployUpgradeConfig < Base
+    @serverUrl = nil
+    @apiKey = nil
+    @upgradeConfig = nil
+
+    @NotificationModes = ['AlwaysShow', 'ShowOnlyMajorMinor', 'NeverShow']
+
+    def initialize(*url_and_api_key)
+      serverUrl, apiKey = get_octopus_creds(url_and_api_key)
+
+      @name = "Octopus Deploy Upgrade Config #{serverUrl}"
+      @runner = Specinfra::Runner
+      @serverUrl = serverUrl
+      @apiKey = apiKey
+
+      # is it still nil?
+      if serverUrl.nil?
+        raise "'serverUrl' was not provided. Unable to connect to Octopus server to validate configuration."
+      end
+      if apiKey.nil?
+        raise "'apiKey' was not provided. Unable to connect to Octopus server to validate configuration."
+      end
+
+      @upgradeConfig = get_upgrade_config_via_api(serverUrl, apiKey)
+    end
+
+    def has_notification_mode?(mode)  # probably needs an enum?
+      false if @upgradeConfig.nil?
+      @upgradeConfig['NotificationMode'] == mode
+    end
+
+    def never_show_notifications?
+      false if @upgradeConfig.nil?
+      has_notification_mode?('NeverShow')
+    end
+
+    def always_show_notifications?
+      false if @upgradeConfig.nil?
+      has_notification_mode?('AlwaysShow')
+    end
+
+    def show_major_minor_notifications?
+      false if @upgradeConfig.nil?
+      has_notification_mode?('ShowOnlyMajorMinor')
+    end
+
+    def include_statistics?
+      false if @upgradeConfig.nil?
+      @upgradeConfig['IncludeStatistics'] == true
+    end
+
+    def allow_checking?
+      false if @upgradeConfig.nil?
+      @upgradeConfig['AllowChecking'] == true
+    end
+  end
+
+  # module-level constructors/entrypoints
+
+  def octopus_deploy_upgrade_config(*url_and_api_key)
+    serverUrl, apiKey = get_octopus_creds(url_and_api_key)
+    OctopusDeployUpgradeConfig.new(serverUrl, apiKey)
+  end
+
+  def octopus_upgrade_config(*url_and_api_key)
+    serverUrl, apiKey = get_octopus_creds(url_and_api_key)
+    octopus_deploy_upgrade_config(serverUrl, apiKey)
+  end
+
+  private
+
+  def get_upgrade_config_via_api(serverUrl, apiKey)
+    smtp = nil
+
+    url = "#{serverUrl}/api/upgradeconfiguration?api-key=#{apiKey}"
```

Comment:
I wonder if we should be following the hypermedia links here? ie, get the root document, then following the `UpgradeConfiguration` link?

Reply:
Is the URI stem ever likely to change, necessitating going back to the links?

Reply:
Potentially. The reason we have the links is so that we can do that.
Also, especially for this one... It shouldn't be it's own endpoint anyway - it should go via the standard configuration api instead...

Reply:
Okey dokey. Change inbound.